lib/tree.rb

```diff
@@ -16,42 +16,111 @@ def initialize
-  # Space Complexity: 
+  # Time Complexity: O(n)
+  # Space Complexity: O(n)
   def add(key, value)
-    raise NotImplementedError
+    @root = add_helper(@root, key, value)
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def add_helper(current_node, key, value)
+    return TreeNode.new(key, value) if current_node.nil?
+
+    if key < current_node.key
+      current_node.left = add_helper(current_node.left, key, value)
+    else
+      current_node.right = add_helper(current_node.right, key, value)
+    end
+    return current_node
+  end 
+
+  # Time Complexity: O(n)
+  # Space Complexity: O(n)
   def find(key)
-    raise NotImplementedError
+    return nil if @root.nil?
+    return find_helper(@root, key)
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def find_helper(current_node, key)
+    return current_node.value if current_node.key == key 
+
+    if key < current_node.key 
+      find_helper(current_node.left, key)
+    else 
+      find_helper(current_node.right, key)
+    end 
+  end 
+
+  # For traversals: Space required is proportional to height of tree which is equal to the number of nodes in the tree in the worst case (skewed trees)
+  
+  # Time Complexity: O(n)
+  # Space Complexity: O(n) or O(h) because the max # of memory addresses on the stack at the same time is the same as the tree height
   def inorder
-    raise NotImplementedError
+    return inorder_helper(@root, [])
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def inorder_helper(current_node, list)
+    # left, root, right
+    return list if current_node.nil?
+    # recursion; left side first
+    inorder_helper(current_node.left, list)
+    list << {key: current_node.key, value: current_node.value}
+    inorder_helper(current_node.right, list)
+    return list
+  end 
+
+  # Time Complexity: O(n)
+  # Space Complexity: O(n)
   def preorder
-    raise NotImplementedError
+    return preorder_helper(@root, [])
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def preorder_helper(current_node, list)
+    # root, left, right
+    return list if current_node.nil?
+    list << {key: current_node.key, value: current_node.value}
+    preorder_helper(current_node.left, list)
+    preorder_helper(current_node.right, list)
+    return list 
+  end 
+
+  # Time Complexity: O(n)
+  # Space Complexity: O(n)
   def postorder
-    raise NotImplementedError
+    return postorder_helper(@root, [])
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def postorder_helper(current_node, list)
+    # left, right, root 
+    return list if current_node.nil?
+
+    postorder_helper(current_node.left, list)
+    postorder_helper(current_node.right, list)
+    list << {key: current_node.key, value: current_node.value}
+
+  end 
+
+  # Time Complexity: 0(n)
+  # Space Complexity: O(n)
+  #Height is distance from the best accessible leaf node 
+  #Height of a tree would be equal to max depth of any node in the tree 
   def height
```

Comment:
👍 
The space complexity is O(log n) if the tree is balanced and O(n) if it's unbalanced.